### PR TITLE
(PUP-4034) AIO - CodeDir / Hiera Changes

### PR DESCRIPTION
**Pull request by ferventcoder**
Create the code directory and copy the default `hiera.yaml` file into
that directory. Also create the `hieradata` directory as a subfolder of
code.

Without these changes, pathing will not match with AIO on different
platforms.

### CODE REVIEW

File: wix/puppet.wxs [hunk 1]
```diff
@@ -350,11 +350,21 @@
...
+                <CreateFolder />
+              </Component>
+            </Directory>
+          </Directory>
           <Directory Id="HieraAppData" Name="hiera">
             <Directory Id="HieraEtcDir" Name="etc">
               <Component Id="HieraEtcDir" Permanent="yes" Guid="88DE0215-554E-44CC-A8EB-CD8C030B1695">
                 <CreateFolder />
-                <File Id="HieraConf" KeyPath="yes" Source="stagedir\hiera\ext\hiera.yaml" />
               </Component>
```
Comment (ferventcoder) [inline wix/puppet.wxs]: The hiera / etc and var directories are still left behind. 

Should I remove this tree structure as well?

```
- hiera
 |
 |- etc
 |- var
```

File: wix/puppet.wxs [hunk 2]
```diff
@@ -350,15 +350,20 @@
...
             </Directory>
-            <Directory Id="HieraVarDir" Name="var">
-              <Component Id="HieraVarDir" Permanent="yes" Guid="01C15CC4-74D1-4E33-887B-E25CA4A8EFB6">
+          </Directory>
+          <Directory Id="HieraAppData" Name="hiera">
+            <Directory Id="HieraEtcDir" Name="etc">
+              <Component Id="HieraEtcDir" Permanent="yes" Guid="88DE0215-554E-44CC-A8EB-CD8C030B1695">
```
Comment (ferventcoder) [inline wix/puppet.wxs]: Does this directory go away as well?

Comment (joshcooper) [inline wix/puppet.wxs]: Definitely `...\PuppetLabs\hiera\etc` can go away, and likely `hiera` as nothing is being stored there any more /cc @haus @hlindberg

Comment (haus) [inline wix/puppet.wxs]: :+1:

Comment (ferventcoder) [inline wix/puppet.wxs]: Removed.

File: wix/puppet.wxs [hunk 3]
```diff
@@ -752,8 +750,8 @@
       <ComponentRef Id="RegistryEntries" />
       <ComponentRef Id="RegistryEntriesArchitectureDependent" />
       <ComponentRef Id="PuppetVarDir" />
-      <ComponentRef Id="HieraVarDir" />
-      <ComponentRef Id="HieraEtcDir" />
+      <ComponentRef Id="CodeDir" />
+      <ComponentRef Id="HieraDataDir" />
```
Comment (joshcooper) [inline wix/puppet.wxs]: Previously we pulled in `HieraVarDir` and `HieraEtcDir`, but not the parent `HieraAppData`. But now we do `CodeDir`, does that matter?

:+1: about creating hiera related directories as `Permanent="yes"`

Aside from the one question, looks good.

Comment (ferventcoder) [inline wix/puppet.wxs]: The parent was created because of the other two components. It wasn't explicitly created.

**Comment by joshcooper** [comment]
The MSI is not creating the `..\PuppetLabs\puppet\cache` directory, but since this is a puppet directory setting, puppet will create it when it runs. So I think it's ok.

:shipit:

**Comment by ferventcoder** [comment]
Fair enough, wanted to be sure there wasn't a permissions kind of thing that needed to be set on the directory.

**Comment by joshcooper** [comment]
Also I noticed there is a `PuppetLabs\puppet\code` but that's because https://github.com/puppetlabs/puppet/pull/3646 hasn't been merged yet. Also https://github.com/puppetlabs/hiera/pull/255 hasn't been merged so the installed hiera.yaml is the old version.

**Comment by Iristyle** [comment]
So @joshcooper I see that your prior comment on puppetlabs/puppet#3646 at https://github.com/puppetlabs/puppet/pull/3646#discussion_r25361774 contradicts your comment above at https://github.com/puppetlabs/puppet_for_the_win/pull/98#issuecomment-76039988 with respect to both `cache` and `var` directories needing to be created by the installer. 

`var` is created already based on https://github.com/puppetlabs/puppet_for_the_win/blob/master/wix/puppet.wxs#L344-L352

```xml
<!-- This PuppetVarDir directory is necessary to fix #12576 otherwise
             puppet cannot create the directories it needs since the parent
             does not exist. -->
<Directory Id="PuppetVarDir" Name="var">
  <Component Id="PuppetVarDir" Permanent="yes" Guid="B95A17F3-CF5E-4EC7-859E-F10C0965645F">
    <CreateFolder />
  </Component>
</Directory>
```

Note the redmine bug (that I couldn't find at https://projects.puppetlabs.com/issues/12576) noting that the installer should be creating the `:vardir`

My assumption would be that we should continue to follow that pattern?

**Comment by Iristyle** [comment]
Ok, I built an MSI based on a combo of this PR and https://github.com/puppetlabs/puppet/pull/3646 to verify the MSI changes with the `run_mode` changes.

After I  installed puppet, I ran a `puppet agent -td`, and as you stated in https://github.com/puppetlabs/puppet_for_the_win/pull/98#issuecomment-76039988 Puppet will create its `:vardir` which is now `C:/ProgramData/PuppetLabs/puppet/cache`

The end result was exactly what we  hoped for:


```
C:\Program Files\Puppet Labs\Puppet\bin>puppet agent --configprint vardir
C:/ProgramData/PuppetLabs/puppet/cache

C:\Program Files\Puppet Labs\Puppet\bin>puppet agent --configprint logdir
C:/ProgramData/PuppetLabs/puppet/var/log

C:\Program Files\Puppet Labs\Puppet\bin>puppet agent --configprint confdir
C:/ProgramData/PuppetLabs/puppet/etc

C:\Program Files\Puppet Labs\Puppet\bin>puppet agent --configprint codedir
C:/ProgramData/PuppetLabs/code

C:\Program Files\Puppet Labs\Puppet\bin>puppet agent --configprint rundir
C:/ProgramData/PuppetLabs/puppet/var/run

C:\Program Files\Puppet Labs\Puppet\bin>dir \ProgramData\PuppetLabs\puppet\cache

 Volume in drive C is Windows 2008
 Volume Serial Number is D8AA-E11A

 Directory of C:\ProgramData\PuppetLabs\puppet\cache

02/25/2015  03:32 PM    <DIR>          .
02/25/2015  03:32 PM    <DIR>          ..
02/25/2015  03:32 PM    <DIR>          clientbucket
02/25/2015  03:32 PM    <DIR>          client_data
02/25/2015  03:32 PM    <DIR>          client_yaml
02/25/2015  03:32 PM    <DIR>          facts.d
02/25/2015  03:32 PM    <DIR>          lib
02/25/2015  03:32 PM    <DIR>          state
               0 File(s)              0 bytes
               8 Dir(s)   2,197,774,336 bytes free

C:\Program Files\Puppet Labs\Puppet\bin>dir \ProgramData\PuppetLabs\code
 Volume in drive C is Windows 2008
 Volume Serial Number is D8AA-E11A

 Directory of C:\ProgramData\PuppetLabs\code

02/25/2015  03:31 PM    <DIR>          .
02/25/2015  03:31 PM    <DIR>          ..
02/25/2015  03:28 PM               329 hiera.yaml
02/25/2015  03:31 PM    <DIR>          hieradata
               1 File(s)            329 bytes
               3 Dir(s)   2,197,774,336 bytes free
```

**Comment by ferventcoder** [comment]
\o/

**Pull request merged**